### PR TITLE
Add length param to ReadableResourceStream::read

### PR DESCRIPTION
**Pull request by trowski**
We can then add this param to the `Socket` interface in amphp/socket.

### CODE REVIEW

File: src/ReadableResourceStream.php [hunk 1]
```diff
@@ -112,12 +115,17 @@ public function read(?Cancellation $cancellation = null): ?string
...
+            throw new \ValueError('The chunk length must be a positive integer');
+        }
+
         if (\feof($this->resource)) {
             $this->free();
 
             return null;
         }
 
+        $this->chunkSize = $length;
```
Comment (kelunik) [inline src/ReadableResourceStream.php:128]: I think it shouldn't modify the `chunkSize` default for this stream?

File: src/ReadableResourceStream.php [hunk 2]
```diff
@@ -169,11 +177,6 @@ public function getResource()
         return $this->resource;
     }
 
-    public function setChunkSize(int $chunkSize): void
-    {
-        $this->chunkSize = $chunkSize;
-    }
-
```
Comment (kelunik) [inline src/ReadableResourceStream.php:176]: I'd keep this.

File: src/ReadableResourceStream.php [hunk 3]
```diff
@@ -100,7 +100,10 @@ public function __construct($stream, int $chunkSize = self::DEFAULT_CHUNK_SIZE)
         };
     }
 
-    public function read(?Cancellation $cancellation = null): ?string
+    /**
+     * @param positive-int $length
+     */
+    public function read(?Cancellation $cancellation = null, int $length = self::DEFAULT_CHUNK_SIZE): ?string
```
Comment (kelunik) [inline src/ReadableResourceStream.php:106]: I somehow dislike that the length is behind the cancellation, but there's nothing we can do about that.

File: src/ReadableResourceStream.php (hunk 1 above)
Comment (trowski) [reply src/ReadableResourceStream.php:128]: There is no default size, this is meant to be internal to the stream now. I forgot to remove it from the constructor.

File: src/ReadableResourceStream.php (hunk 2 above)
Comment (trowski) [reply src/ReadableResourceStream.php:176]: It serves no purpose if read has a length param.

File: src/ReadableResourceStream.php (hunk 3 above)
Comment (trowski) [reply src/ReadableResourceStream.php:106]: Luckily we have named params now, so `$stream->read(length: 42)` seems acceptable.

File: src/ReadableResourceStream.php (hunk 1 above)
Comment (kelunik) [reply src/ReadableResourceStream.php:127]: Every reader following the interface should use the chunk size configured in the constructor IMO.

Comment (trowski) [reply src/ReadableResourceStream.php:127]: What interface? We generally don't expose either the constructor or the ability to call `setChunkSize`.

Comment (kelunik) [reply src/ReadableResourceStream.php:127]: The `ReadableStream` interface, e.g. `Amp\ByteStream\pipe` won't have a configurable chunk size anymore.

Comment (trowski) [reply src/ReadableResourceStream.php:127]: Most cases where `ReadableResourceStream` is used provides no opportunity to change the chunk size. `ReadableStream` has no concept of chunk size, it just returns a string of arbitrary size.

File: examples/benchmark-throughput.php [hunk 4]
```diff
@@ -26,7 +26,7 @@
 $of = \preg_replace('(^/dev/fd/)', 'php://fd/', $of);
 
 $stderr = new WritableResourceStream(STDERR);
-$in = new ReadableResourceStream(\fopen($if, 'rb'), 65536 /* Default size used by React to allow comparisons */);
+$in = new ReadableResourceStream(\fopen($if, 'rb'));
```
Comment (kelunik) [inline examples/benchmark-throughput.php:29]: This can be reverted, no?

Comment (trowski) [reply examples/benchmark-throughput.php:29]: Oops, yes, good catch.

**Review by kelunik — APPROVED** [review]

**Pull request merged**